### ISSUE TITLE
Inconsistent behavior between search and sidebar when modeline overrides language specified by extension

### ISSUE THREAD

**Issue by varungandhi-src** [sourcegraph/sourcegraph#60580]
Related items:
- https://github.com/sourcegraph/sourcegraph/pull/59234 - I made the code first check the extension to avoid fetching the file contents (because that can be slow)
- https://github.com/sourcegraph/sourcegraph/issues/60341

[Search query in screenshot](https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/llvm/llvm-project%24+lang:C%2B%2B+file:%5C.c%24&patternType=keyword&sm=0&feat=search-content-based-lang-detection), [file in screenshot](https://sourcegraph.com/github.com/llvm/llvm-project@3af5c98200e0b1268f755c3f289be4f73aac4214/-/blob/llvm/bindings/ocaml/llvm/llvm_ocaml.c)

![](https://github.com/sourcegraph/sourcegraph/assets/93103176/93e13a8f-c631-465b-b397-6008ba1ac002)

One possible fix here would be that once we have a high performance API for fetching file contents ([Slack link](https://sourcegraph.slack.com/archives/C05EMJM2SLR/p1704126731706369?thread_ts=1703827325.504129&cid=C05EMJM2SLR) -- TODO: Replace with GitHub issue link once that is filed), we can also do content-based language detection for the icons and ensure consistent behavior.

As a quick spot check, I went through ~10 different files which have a C++ modeline but a C extension, and all of them seem to actually be valid C and not using any C++ features. So I'm not sure why they got marked as C++ with the modeline. 🤷🏽

**Comment by jtibshirani** [comment]
As a tactical fix, I wonder if we should remove the modeline detection in our `go-enry` wrapper? I'd be really curious to understand how much it helps in practice.

**Comment by varungandhi-src** [comment]
I don't think we should remove the modeline detection because people use that a bunch in practice. Here's an [example query](https://sourcegraph.com/search?q=context:global+-*-+python+-*-+%28NOT+file:.py%24%29+%28NOT+file:.pyx%24%29+-content:%23%21/usr/bin/env&patternType=standard&sm=0) for Python.
